presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchange.java

```diff
@@ -0,0 +1,584 @@
+
+            return page;
+        }
+
+        public ListenableFuture<?> waitForReading()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                if (finishing || !buffer.isEmpty()) {
+                    return NOT_BLOCKED;
+                }
+            }
+            return readWaiter.waitFor();
+        }
+
+        public boolean isFinished()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                return finishing && buffer.isEmpty();
+            }
+        }
+
+        public void finish()
+        {
+            checkNotHoldsLock(this);
+
+            synchronized (this) {
+                if (finishing) {
+                    return;
+                }
+                finishing = true;
+            }
+
+            // notify reader outside of lock
+            readWaiter.finishAndNotifyWaiters();
+
+            // notify finish listener outside of lock
+            if (isFinished()) {
+                onFinish.accept(this);
+            }
+        }
+
+        public void close()
+        {
+            checkNotHoldsLock(this);
+
+            List<PageReference> remainingPages;
+            synchronized (this) {
+                finishing = true;
+
+                remainingPages = ImmutableList.copyOf(buffer);
+                buffer.clear();
+                bufferedBytes.set(0);
+            }
+
+            // free the remaining pages
+            remainingPages.forEach(PageReference::removePage);
+
+            // notify reader outside of lock
+            readWaiter.finishAndNotifyWaiters();
+
+            // notify finish listener outside of lock
+            onFinish.accept(this);
+        }
+    }
+
+    public static class LocalExchangeBufferInfo
+            implements Mergeable<LocalExchangeBufferInfo>
+    {
+        private final long bufferedBytes;
+        private final int bufferedPages;
+
+        @JsonCreator
+        public LocalExchangeBufferInfo(
+                @JsonProperty("bufferedBytes") long bufferedBytes,
+                @JsonProperty("bufferedPages") int bufferedPages)
+        {
+            this.bufferedBytes = bufferedBytes;
+            this.bufferedPages = bufferedPages;
+        }
+
+        @JsonProperty
+        public long getBufferedBytes()
+        {
+            return bufferedBytes;
+        }
+
+        @JsonProperty
+        public int getBufferedPages()
+        {
+            return bufferedPages;
+        }
+
+        @Override
+        public LocalExchangeBufferInfo mergeWith(LocalExchangeBufferInfo other)
+        {
+            return new LocalExchangeBufferInfo(bufferedBytes + other.getBufferedBytes(), bufferedPages + other.getBufferedPages());
+        }
+    }
+
+    @ThreadSafe
+    public static class LocalExchangeSinkFactory
+            implements Closeable
+    {
+        private final LocalExchange exchange;
+
+        @GuardedBy("this")
+        private boolean closed;
```

Comment:
Assuming the complexity of this class does not increase as a result of further code reviews, `AtomicBoolean` seems to be easier to use and produce cleaner code than `sync(this)` and a `boolean`.